model/device.go

```diff
@@ -78,6 +132,48 @@ func (da DeviceAttribute) Validate() error {
 	)
 }
 
+func (da DeviceAttribute) Equal(e DeviceAttribute) bool {
+	if (da.Value == nil || e.Value == nil) && !(da.Value == nil && e.Value == nil) {
+		return false
+	}
+	rVal1 := reflect.ValueOf(da.Value)
+	rVal2 := reflect.ValueOf(e.Value)
+	if rVal1.Kind() != rVal2.Kind() {
+		return false
+	}
+	if rVal1.Kind() == reflect.Slice {
+		if rVal1.Len() != rVal2.Len() {
+			return false
+		}
+		for i := 0; i < rVal1.Len(); i++ {
+			if rVal1.Index(i) != rVal2.Index(i) {
```

Comment:
Comparing two different `reflect.Value` will always evaluate to false. You need to use `reflect.Value.Equal` to compare the underlying values.
```suggestion
			if !rVal1.Comparable() || !rVal1.Index(i).Equal(rVal2.Index(i)) {
```

Reply:
I changed it, because it panicked. I will check that, we have tests cases around it.

Reply:
wow. I missed Comparable. will you believe it? I was just about to complain about lack of comparable in go, and here it is. thanks. and sorry.

Reply:
> Comparing two different reflect.Value will always evaluate to false. You need to use reflect.Value.Equal to compare the underlying values.

this is only true for e.g.: `[]int{1,2,3}` and `[3]int{1,2,3}` -- but this (for supported types) I managed to get.
And I remember the Comparable now, I have seen it, it does not work like that for all the cases we are testing:
```
model/device.go:149:14: rVal1.Comparable undefined (type reflect.Value has no field or method Comparable)model/device.go:149:46: rVal1.Index(i).Equal undefined (type reflect.Value has no field or method Equal)
```

And I also checked that the `rVal1.Index(i) != rVal2.Index(i)` does fill the hole.

I honestly have no other idea how to pass all the tests we are having (I had added Equal call to all of then before I asked for the second (I think) review). Are we really using all those attributes from the units in the field?

Reply:
What about just comparing the values by `reflect.DeepEqual` since that is what you're essentially doing?

Reply:
> What about just comparing the values by `reflect.DeepEqual` since that is what you're essentially doing?

I tried DeepEqual -- it does not really work for our case. even equal elements are not considered equal.

Reply:
@alfrunes we have to move on with this :) I believe it should be ok for our use-cases since 99% of people have primitive types and flat attributes. do you want me to re-visit the DeepEqual and try to come up with something?

Reply:
Comparing `reflect.Value`s does not work. Try this test case:
```go
	assert.True(t,
		DeviceAttribute{Name: "foo", Value: []interface{}{123.0}}.Equal(
			DeviceAttribute{Name: "foo", Value: []interface{}{123.0}},
		))
```
The reason your test passes is because you're comparing the same slice reference.

Reply:
this is a killer :) ok I will revisit this. there has to be a better way to do it.

Reply:
man, I swear I was fooling around DeepEqual, and I have no idea, but now all is fine. what do you think? green light?